### PR TITLE
Remove nalgebra in favour of a handmade function

### LINKED ISSUE #241: Remove nalgebra
Currently, `nalgebra` is used just for [one method](https://github.com/tomaka/rodio/blob/master/src/source/spatial.rs#L42..L55), but it's a very heavy dependency:

```
`-- nalgebra v0.18.0
    |-- alga v0.9.1
    |   |-- approx v0.3.2
    |   |   `-- num-traits v0.2.8
    |   |       [build-dependencies]
    |   |       `-- autocfg v0.1.5
    |   |-- libm v0.1.4
    |   |-- num-complex v0.2.3
    |   |   `-- num-traits v0.2.8 (*)
    |   |   [build-dependencies]
    |   |   `-- autocfg v0.1.5 (*)
    |   `-- num-traits v0.2.8 (*)
    |-- approx v0.3.2 (*)
    |-- generic-array v0.12.3
    |   `-- typenum v1.10.0
    |-- matrixmultiply v0.2.2
    |   `-- rawpointer v0.1.0
    |-- num-complex v0.2.3 (*)
    |-- num-traits v0.2.8 (*)
    |-- rand v0.6.5
    |   |-- libc v0.2.60 (*)
    |   |-- rand_chacha v0.1.1
    |   |   `-- rand_core v0.3.1
    |   |       `-- rand_core v0.4.0
    |   |   [build-dependencies]
    |   |   `-- autocfg v0.1.5 (*)
    |   |-- rand_core v0.4.0 (*)
    |   |-- rand_hc v0.1.0
    |   |   `-- rand_core v0.3.1 (*)
    |   |-- rand_isaac v0.1.1
    |   |   `-- rand_core v0.3.1 (*)
    |   |-- rand_jitter v0.1.4
    |   |   `-- rand_core v0.4.0 (*)
    |   |-- rand_os v0.1.3
    |   |   |-- libc v0.2.60 (*)
    |   |   `-- rand_core v0.4.0 (*)
    |   |-- rand_pcg v0.1.2
    |   |   `-- rand_core v0.4.0 (*)
    |   |   [build-dependencies]
    |   |   `-- autocfg v0.1.5 (*)
    |   `-- rand_xorshift v0.1.1
    |       `-- rand_core v0.3.1 (*)
    |   [build-dependencies]
    |   `-- autocfg v0.1.5 (*)
    `-- typenum v1.10.0 (*)
```

```
  Time Crate
29.16s nalgebra
 7.79s cc
 6.57s lewton
 3.78s libm
 3.19s pkg_config
 3.14s rodio
 2.96s rand
 2.88s alga
 2.41s matrixmultiply
 2.12s cpal
 1.84s num_traits
 1.83s libc
 1.74s byteorder
 1.52s ogg
 1.47s autocfg
 1.25s build_script_main
 1.16s claxon
 1.09s num_complex
 0.91s slice_deque
 0.90s typenum
 0.87s minimp3_sys
 0.78s alsa_sys
 0.67s rand_os
 0.61s rand_isaac
 0.57s generic_array
 0.54s rand_chacha
 0.53s rand_hc
 0.52s hound
 0.47s rand_core
 0.42s smallvec
 0.38s rand_jitter
 0.35s rand_pcg
 0.20s rand_xorshift
 0.18s approx
 0.15s minimp3
 0.08s lazy_static
 0.05s rawpointer
 0.04s rand_core
```

I'm not sure how hard is to reimplement this method manually though.

### PR DESCRIPTION
**Pull request by est31**
Fixes #241

### CODE REVIEW

File: src/source/spatial.rs [hunk 1]
```diff
@@ -16,6 +15,13 @@ where
     input: ChannelVolume<I>,
 }
 
+fn dist(a: [f32; 3], b: [f32; 3]) -> f32 {
+    a.iter().zip(b.iter())
+        .map(|(a, b)| (a - b) * (a - b))
+        .sum::<f32>()
+        .sqrt()
```
Comment (OvermindDL1) [inline src/source/spatial.rs]: Technically you can save a lot of work and processing time by not using a `.sqrt()`, and adjusting the `set_positions` function... (see next message)

File: src/source/spatial.rs [hunk 2]
```diff
@@ -39,12 +45,9 @@ where
...
-        let right_ear = Point3::from(right_ear);
-        let left_distance = (left_ear - emitter_position).magnitude();
-        let right_distance = (right_ear - emitter_position).magnitude();
-        let max_diff = (left_ear - right_ear).magnitude();
+        let left_distance = dist(left_ear, emitter_pos);
+        let right_distance = dist(right_ear, emitter_pos);
+        let max_diff = dist(left_ear, right_ear);
         let left_diff_modifier = ((left_distance - right_distance) / max_diff + 1.0) / 4.0 + 0.5;
         let right_diff_modifier = ((right_distance - left_distance) / max_diff + 1.0) / 4.0 + 0.5;
         let left_dist_modifier = (1.0 / left_distance.powi(2)).min(1.0);
```
Comment (OvermindDL1) [inline src/source/spatial.rs]: And adjusting these `left_dist_modifier` and right_dist_modifier` lines.

Comment (est31) [reply src/source/spatial.rs]: Good idea, thanks!

File: src/source/spatial.rs [hunk 3]
```diff
@@ -45,13 +44,15 @@ where
...
-        let max_diff = dist(left_ear, right_ear);
-        let left_diff_modifier = ((left_distance - right_distance) / max_diff + 1.0) / 4.0 + 0.5;
-        let right_diff_modifier = ((right_distance - left_distance) / max_diff + 1.0) / 4.0 + 0.5;
-        let left_dist_modifier = (1.0 / left_distance.powi(2)).min(1.0);
-        let right_dist_modifier = (1.0 / right_distance.powi(2)).min(1.0);
+        let left_dist_sq = dist_sq(left_ear, emitter_pos);
+        let right_dist_sq = dist_sq(right_ear, emitter_pos);
+        let max_diff = dist_sq(left_ear, right_ear).sqrt();
+        let left_dist = left_dist_sq.sqrt();
+        let right_dist = right_dist_sq.sqrt();
```
Comment (OvermindDL1) [inline src/source/spatial.rs]: Do you still need the sqrt's at all, or could the math below be adjusted to not require them.

Comment (est31) [reply src/source/spatial.rs]: I've tried to think about alternatives but didn't have any ideas. I think there is no other way.

Comment (est31) [reply src/source/spatial.rs]: If you know of an alternative @overminddl1 please file a PR and we can discuss it.

**Pull request merged**